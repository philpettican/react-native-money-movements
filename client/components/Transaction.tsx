import React from 'react';
import { Image, StyleSheet, Text, View, ViewStyle } from 'react-native';
import { SvgUri } from 'react-native-svg';
import { ListItem, ListItemProps } from 'react-native-elements';
import { MaterialCommunityIcons } from '@expo/vector-icons';
import moment from 'moment';

import TransactionModel, { ITransaction } from '../model/Transaction';

export type TransactionProps = ListItemProps & {
    transaction: ITransaction
};

export type TransactionIconProps = {
    containerStyle?: ViewStyle,
    uri?: string;
    width?: number;
    height?: number;
};

export const TransactionIcon = (props: TransactionIconProps) => {
    const { containerStyle, uri, width, height } = props;

    if (!uri) {
        return null;
    }

    const isSvg = uri.endsWith('.svg');

    return (
        <View style={containerStyle}>
            {isSvg && <SvgUri testID="svg" uri={uri} width={width || styles.imageStyle.width} height={height || styles.imageStyle.height}></SvgUri>}
            {!isSvg && <Image testID="image" source={{uri}} style={StyleSheet.flatten([styles.imageStyle, width ? { width } : null, height ? { height } : null])}></Image>}
        </View>
    )
};

export const CarbonFootPrintIcon = () => {
    return (
        <MaterialCommunityIcons name="foot-print" size={25}></MaterialCommunityIcons>
    );
};

const Transaction = (props: TransactionProps) => {
    const { onPress } = props;
    const transaction = new TransactionModel(props.transaction);
    const { name, icon, created_at, carbon_footprint } = transaction;

    return (
        <ListItem bottomDivider onPress={onPress}>
            <TransactionIcon uri={icon}></TransactionIcon>
            <ListItem.Content>
                <ListItem.Title>
                    {name}
                </ListItem.Title>
                <ListItem.Subtitle>
                    {moment(created_at).format('LLL')}
                </ListItem.Subtitle>
            </ListItem.Content>
            <View style={styles.rightElementContainerStyle}>
                <Text>{transaction.amountWithCurrencySymbol}</Text>
                <View style={styles.carbonFootPrintContainerStyle}>
                    <Text>{carbon_footprint}</Text>
                    <CarbonFootPrintIcon></CarbonFootPrintIcon>
                </View>
            </View>
            {!!onPress && <ListItem.Chevron></ListItem.Chevron>}
        </ListItem>
    )
};

const styles = StyleSheet.create({
    imageStyle: {
        width: 50,
        height: 50,
    },
    rightElementContainerStyle: {
        width: 100,
        alignItems: 'flex-end',
    },
    carbonFootPrintContainerStyle: {
        flexDirection: 'row',
        alignItems: 'center',
    },
});

export default Transaction;
