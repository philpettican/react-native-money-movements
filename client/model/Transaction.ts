import getSymbolFromCurrency from "currency-symbol-map";

export interface ITransaction {
    name: string,
    status: string,
    category: string,
    carbon_footprint: number,
    fees: number,
    type: string,
    amount: number,
    currency: string,
    created_at: string,
    icon: string,
    brand_partner: string
};

export default class Transaction implements ITransaction {
    name: string;
    status: string;
    category: string;
    carbon_footprint: number;
    fees: number;
    type: string;
    amount: number;
    currency: string;
    created_at: string;
    icon: string;
    brand_partner: string;

    constructor(props: ITransaction) {
        this.name = props.name;
        this.status = props.status;
        this.category = props.category;
        this.carbon_footprint = props.carbon_footprint;
        this.fees = props.fees;
        this.type = props.type;
        this.amount = props.amount;
        this.currency = props.currency;
        this.created_at = props.created_at;
        this.icon = props.icon;
        this.brand_partner = props.brand_partner;
    }

    get currencySymbol() {
        return getSymbolFromCurrency(this.currency);
    }

    get amountWithCurrencySymbol() {
        return `${this.currencySymbol}${this.amount.toFixed(2)}`;
    }

}