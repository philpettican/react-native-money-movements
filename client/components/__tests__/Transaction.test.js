import * as React from 'react';
import { render } from '@testing-library/react-native';

import Transaction from '../Transaction';

const transactionData = {
    name: "Tuition Fees",
    status: "completed",
    category: "education",
    carbon_footprint: 25,
    fees: 0,
    type: "card",
    amount: 2195,
    currency: "GBP",
    created_at: "2021-04-22 02:15",
    icon: "https://novus-app-styleguide.s3.eu-west-1.amazonaws.com/banking_icons/education_small.svg",
    brand_partner: "FALSE"
};

it(`renders correctly`, () => {
  const { getAllByText, toJSON } = render(<Transaction transaction={transactionData}></Transaction>);
  const tree = toJSON();
  expect(tree).toMatchSnapshot();
  expect(getAllByText(transactionData.name).length).toBe(1);
  expect(getAllByText(`Â£${transactionData.amount.toFixed(2)}`).length).toBe(1);
});

it('renders svg if icon extension is .svg', () => {
  const { getByTestId } = render(<Transaction transaction={transactionData}></Transaction>);
  expect(getByTestId('svg')).toBeDefined();
});

it('renders image if icon extension is not .svg', () => {
  const { getByTestId } = render(<Transaction transaction={{...transactionData, icon: transactionData.icon.replace('.svg', '.png')}}></Transaction>);
  expect(getByTestId('image')).toBeDefined();
});