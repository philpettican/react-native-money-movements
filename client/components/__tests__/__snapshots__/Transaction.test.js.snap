// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#ffffff",
        "borderBottomWidth": 0.5,
        "borderColor": "#bcbbc1",
        "flexDirection": "row",
        "padding": 14,
      }
    }
    testID="padView"
  >
    <View>
      <SvgUri
        height={50}
        testID="svg"
        uri="https://novus-app-styleguide.s3.eu-west-1.amazonaws.com/banking_icons/education_small.svg"
        width={50}
      />
    </View>
    <View
      style={
        Object {
          "paddingLeft": 16,
        }
      }
    />
    <View
      style={
        Array [
          Object {
            "alignItems": "flex-start",
            "flex": 1,
            "justifyContent": "center",
          },
          undefined,
        ]
      }
      theme={
        Object {
          "colors": Object {
            "black": "#242424",
            "disabled": "hsl(208, 8%, 90%)",
            "divider": "#bcbbc1",
            "error": "#ff190c",
            "grey0": "#393e42",
            "grey1": "#43484d",
            "grey2": "#5e6977",
            "grey3": "#86939e",
            "grey4": "#bdc6cf",
            "grey5": "#e1e8ee",
            "greyOutline": "#bbb",
            "platform": Object {
              "android": Object {
                "error": "#f44336",
                "grey": "rgba(0, 0, 0, 0.54)",
                "primary": "#2196f3",
                "searchBg": "#dcdce1",
                "secondary": "#9C27B0",
                "success": "#4caf50",
                "warning": "#ffeb3b",
              },
              "default": Object {
                "error": "#ff3b30",
                "grey": "#7d7d7d",
                "primary": "#007aff",
                "searchBg": "#dcdce1",
                "secondary": "#5856d6",
                "success": "#4cd964",
                "warning": "#ffcc00",
              },
              "ios": Object {
                "error": "#ff3b30",
                "grey": "#7d7d7d",
                "primary": "#007aff",
                "searchBg": "#dcdce1",
                "secondary": "#5856d6",
                "success": "#4cd964",
                "warning": "#ffcc00",
              },
              "web": Object {
                "error": "#ff190c",
                "grey": "#393e42",
                "primary": "#2089dc",
                "searchBg": "#303337",
                "secondary": "#ca71eb",
                "success": "#52c41a",
                "warning": "#faad14",
              },
            },
            "primary": "#2089dc",
            "searchBg": "#303337",
            "secondary": "#ca71eb",
            "success": "#52c41a",
            "warning": "#faad14",
            "white": "#ffffff",
          },
        }
      }
    >
      <Text
        style={
          Object {
            "backgroundColor": "transparent",
            "color": "#242424",
            "fontSize": 17,
          }
        }
        testID="listItemTitle"
      >
        Tuition Fees
      </Text>
      <Text
        style={
          Object {
            "backgroundColor": "transparent",
            "color": "#242424",
            "fontSize": 15,
          }
        }
        testID="listItemTitle"
      >
        April 22, 2021 2:15 AM
      </Text>
    </View>
    <View
      style={
        Object {
          "paddingLeft": 16,
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "flex-end",
          "width": 100,
        }
      }
    >
      <Text>
        Â£2195.00
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Text>
          25
        </Text>
      </View>
    </View>
  </View>
</View>
`;
